scripts/morty/data.py

```diff
@@ -0,0 +1,314 @@
+                except KeyError:
+                    continue
+        micro_net_xs = net_xs * 1e-24
+        macro_fiss_xs = fiss_xs * self.run_params['fissile_atom_dens_cc']
+        return micro_net_xs, macro_fiss_xs
+
+    def openmc_data_gen(self):
+        """
+        Constructs dictionaries of data based on OpenMC data
+
+        Returns
+        -------
+        data_params : dict
+            key : str
+                Name of data collection
+            value : dict
+                Dictionary of results
+        """
+        import openmc.deplete
+        lams = {}
+        FYs = {}
+        loss_rates = {}
+        decay_frac = {}
+        tracked_nucs = {}
+        tracked_element = self.target_element
+
+        chain = openmc.deplete.Chain.from_xml(self.chain_path)
+        yield_fracs = chain[self.fissile_nuclide].yield_data
+        cur_target = f'{tracked_element}{self.target_isobar}'
+        target_atomic_number = self.atomic_numbers[tracked_element]
+        i = 0
+        feeds = i - 1
+        skip_feed = []
+        _, fiss_xs = self._get_tot_xs(cur_target)
+        while i < self.num_nucs:
+            cur_target = f'{tracked_element}{self.target_isobar}'
+            tracked_nucs[i] = cur_target
+            lams[i] = openmc.data.decay_constant(cur_target)
+            net_xs, _ = self._get_tot_xs(cur_target)
+            FYs[i] = self.flux * fiss_xs * yield_fracs[self.energy][cur_target]
+            loss_rates[i] = net_xs * self.flux
+
+            decays = chain.nuclides[chain.nuclide_dict[cur_target]].decay_modes
+            if i != 0 and len(decays) > 1:
+                for dec in decays:
+                    dec_type, target, ratio = dec
+                    if self.target_isobar not in target:
+                        continue
+                    if target == prev_target:
+                        feed_val = feeds
+                        while feed_val in skip_feed:
+                            feed_val += 1
+                        decay_frac[(i, feed_val)] = ratio
+                        continue
+                    decay_frac[(i, feeds + 2)] = ratio
+
+                    i += 1
+                    if i >= self.num_nucs:
+                        break
+                    tracked_nucs[i] = target
+                    net_xs, _ = self._get_tot_xs(cur_target)
+                    loss_rates[i] = net_xs * self.flux
+                    lams[i] = openmc.data.decay_constant(target)
+                    FYs[i] = self.flux * fiss_xs * \
+                        yield_fracs[self.energy][target]
+                    if "_m" in target:
+                        decay_frac[(i, feeds)] = 1
+                        skip_feed.append(i)
+            else:
+                decay_frac[(i, feeds)] = 1
+            i += 1
+            feeds += 1
+            target_atomic_number -= 1
+            tracked_element = self.atomic_symbols[target_atomic_number]
+            prev_target = cur_target
+
```

Comment:
This is a very long while loop. I recommend pulling out a few sections and making a few smaller, testable functions.